File: general-interview-prep/javascript/relativeSortArray.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
    // keep track of count of each item that is found in arr1 that is also in arr2
    // then keep track of only those items which do not have colocation in both
    // append the sorted partition which was not found in the second array
    // and return concatenated result

    let included = {};
    let unincluded = {};

    arr1.forEach(num => {
        if (arr2.includes(num)) {
            if (included[num] === undefined) {
                included[num] = 1;
            } else {
                included[num]++;
            }
        }
    });

    arr1.forEach(num => {
        if (!arr2.includes(num)) {
            if (unincluded[num] === undefined) {
                unincluded[num] = 1;
            } else {
                unincluded[num]++;
            }
        }
    });

    let firstPartition = [];
    let secondPartition = [];

    for (let i = 0; i < arr2.length; i++) {
        while(included[arr2[i]] !== 0) {
            firstPartition.push(arr2[i]);
            included[arr2[i]]--;
        }
    }

    for (let key in unincluded) {
        while(unincluded[key] !== 0) {
            secondPartition.push(key);
            unincluded[key]--;
        }
    }

    secondPartition.sort((a, b) => a - b);
    return [...firstPartition, ...secondPartition];
};

File: general-interview-prep/golang/relativeSortArray.go
package main

import (
	"fmt",
	"sort"
)


func contains(num int, arr []int) bool {
		for i := 0; i < len(arr); i++ {
				if arr[i] == num {
						return true
				}
		}
		
		return false
}
	
func relativeSortArray(arr1 []int, arr2 []int) []int {
		countMap := map[int]int{}
		result := make([]int, 0)
		end := make([]int, 0)

		for _, num := range arr1 {
				if contains(num, arr2) == true {
						countMap[num]++
				} else {
						end = append(end, num)
				}
		}
		
		for _, num := range arr2 {
				count := countMap[num]
				for i := 0; i < count; i++ {
						result = append(result, num)
				}
		}
		
		sort.Ints(end)
		result = append(result, end...)
		
		return result
}

func main() {
	fmt.Println(relativeSortArray([]int{1, 2, 3, 4}, []int{2, 3}))
}